src/mscorlib/shared/System/SpanHelpers.Char.cs

```diff
@@ -84,79 +84,86 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
+                    int unaligned = ((int)Unsafe.AsPointer(ref searchSpace) & (Vector<byte>.Count - 1)) / elementsPerByte;
+                    length = ((Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1));
+                }
+            SequentialScan:
 #endif
-            while ((byte*)nLength >= (byte*)4)
-            {
-                nLength -= 4;
-
-                if (uValue == Unsafe.Add(ref searchSpace, index))
-                    goto Found;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 1))
-                    goto Found1;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 2))
-                    goto Found2;
-                if (uValue == Unsafe.Add(ref searchSpace, index + 3))
-                    goto Found3;
-
-                index += 4;
-            }
+                while (length >= 4)
+                {
+                    length -= 4;
+
+                    if (*pCh == value)
+                        goto Found;
+                    if (*(pCh + 1) == value)
+                        goto Found1;
+                    if (*(pCh + 2) == value)
+                        goto Found2;
+                    if (*(pCh + 3) == value)
+                        goto Found3;
+
+                    pCh += 4;
+                }
 
-            while ((byte*)nLength > (byte*)0)
-            {
-                nLength -= 1;
+                while (length > 0)
+                {
+                    length -= 1;
 
-                if (uValue == Unsafe.Add(ref searchSpace, index))
-                    goto Found;
+                    if (*pCh == value)
+                        goto Found;
 
-                index += 1;
-            }
+                    pCh += 1;
+                }
 #if !netstandard11
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
-            {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<ushort>.Count - 1));
+                if (Vector.IsHardwareAccelerated && pCh < pEndCh)
+                {
+                    length = (int)((pEndCh - pCh) & ~(Vector<ushort>.Count - 1));
 
-                // Get comparison Vector
-                Vector<ushort> vComparison = new Vector<ushort>(value);
+                    // Get comparison Vector
+                    Vector<ushort> vComparison = new Vector<ushort>(value);
 
-                while ((byte*)nLength > (byte*)index)
-                {
-                    var vMatches = Vector.Equals(vComparison, Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref searchSpace, index))));
-                    if (Vector<ushort>.Zero.Equals(vMatches))
+                    while (length > 0)
                     {
-                        index += Vector<ushort>.Count;
-                        continue;
+                        Vector<ushort> vMatches = Vector.Equals(vComparison, Unsafe.ReadUnaligned<Vector<ushort>>(pCh));
```

Comment:
You do not need `Unsafe.ReadUnaligned` anymore because of the value is guaranteed to be aligned thanks to pinning. It can be just a `Unsafe.Read`.

Reply:
`ROSpans<char>` from a string will be aligned, but what about from an arbitrary slice of a char array (or even a slice of a string)?